File: packages/aid-java/src/main/java/org/agentcommunity/aid/Parser.java
package org.agentcommunity.aid;

import java.net.URI;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

public final class Parser {
  private Parser() {}

  public static AidRecord parse(String txtRecord) {
    Map<String, String> raw = parseRawRecord(txtRecord);
    return validateRecord(raw);
  }

  private static Map<String, String> parseRawRecord(String txtRecord) {
    Map<String, String> record = new HashMap<>();
    String[] parts = txtRecord.split(";");
    for (String part : parts) {
      String pair = part.trim();
      if (pair.isEmpty()) continue;
      int idx = pair.indexOf('=');
      if (idx < 0) {
        throw new AidError("ERR_INVALID_TXT", "Invalid key-value pair: " + pair);
      }
      String key = pair.substring(0, idx).trim().toLowerCase(Locale.ROOT);
      String value = pair.substring(idx + 1).trim();
      if (key.isEmpty() || value.isEmpty()) {
        throw new AidError("ERR_INVALID_TXT", "Empty key or value in pair: " + pair);
      }
      if (record.containsKey(key)) {
        throw new AidError("ERR_INVALID_TXT", "Duplicate key: " + key);
      }
      // Only store known keys; ignore unknown for fwd-compat
      switch (key) {
        case "v":
        case "uri":
        case "u":
        case "proto":
        case "p":
        case "auth":
        case "a":
        case "desc":
        case "s":
        case "docs":
        case "d":
        case "dep":
        case "e":
        case "pka":
        case "k":
        case "kid":
        case "i":
          record.put(key, value);
          break;
        default:
          // ignore unknown
      }
    }
    return record;
  }

  public static AidRecord validateRecord(Map<String, String> raw) {
    // Required fields
    if (!raw.containsKey("v")) {
      throw new AidError("ERR_INVALID_TXT", "Missing required field: v");
    }
    boolean hasUri = raw.containsKey("uri");
    boolean hasU = raw.containsKey("u");
    if (hasUri && hasU) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"uri\" and \"u\"");
    }
    if (!hasUri && !hasU) {
      throw new AidError("ERR_INVALID_TXT", "Missing required field: uri");
    }

    boolean hasProto = raw.containsKey("proto");
    boolean hasP = raw.containsKey("p");
    if (hasProto && hasP) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"proto\" and \"p\" fields");
    }
    if (!hasProto && !hasP) {
      throw new AidError("ERR_INVALID_TXT", "Missing required field: proto (or p)");
    }

    String version = raw.get("v");
    if (!Constants.SPEC_VERSION.equals(version)) {
      throw new AidError(
          "ERR_INVALID_TXT",
          "Unsupported version: " + version + ". Expected: " + Constants.SPEC_VERSION);
    }

    String protoValue = hasProto ? raw.get("proto") : raw.get("p");

    // Validate protocol token: only compare against known constants
    if (!isValidProto(protoValue)) {
      throw new AidError("ERR_UNSUPPORTED_PROTO", "Unsupported protocol: " + protoValue);
    }

    // Auth token
    if (raw.containsKey("auth") && raw.containsKey("a")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"auth\" and \"a\" fields");
    }
    String authVal = raw.containsKey("auth") ? raw.get("auth") : (raw.containsKey("a") ? raw.get("a") : null);
    if (authVal != null && !isValidAuth(authVal)) {
      throw new AidError("ERR_INVALID_TXT", "Invalid auth token: " + authVal);
    }

    // Desc length check (≤ 60 UTF-8 bytes)
    if (raw.containsKey("desc") && raw.containsKey("s")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"desc\" and \"s\" fields");
    }
    String descVal = raw.containsKey("desc") ? raw.get("desc") : (raw.containsKey("s") ? raw.get("s") : null);
    if (descVal != null) {
      int bytes = descVal.getBytes(java.nio.charset.StandardCharsets.UTF_8).length;
      if (bytes > 60) {
        throw new AidError("ERR_INVALID_TXT", "Description field must be ≤ 60 UTF-8 bytes");
      }
    }
    if (raw.containsKey("docs") && raw.containsKey("d")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"docs\" and \"d\" fields");
    }
    String docsVal = raw.containsKey("docs") ? raw.get("docs") : (raw.containsKey("d") ? raw.get("d") : null);
    if (docsVal != null) {
      if (!docsVal.startsWith("https://")) {
        throw new AidError("ERR_INVALID_TXT", "docs MUST be an absolute https:// URL");
      }
      try {
        java.net.URI du = java.net.URI.create(docsVal);
        if (!"https".equals(du.getScheme()) || du.getHost() == null) {
          throw new IllegalArgumentException();
        }
      } catch (Exception e) {
        throw new AidError("ERR_INVALID_TXT", "Invalid docs URL: " + docsVal);
      }
    }
    if (raw.containsKey("dep") && raw.containsKey("e")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"dep\" and \"e\" fields");
    }
    String depVal = raw.containsKey("dep") ? raw.get("dep") : (raw.containsKey("e") ? raw.get("e") : null);
    if (depVal != null) {
      if (!depVal.endsWith("Z")) {
        throw new AidError("ERR_INVALID_TXT", "dep MUST be an ISO 8601 UTC timestamp (e.g., 2026-01-01T00:00:00Z)");
      }
      try {
        java.time.Instant dep = java.time.Instant.parse(depVal);
        if (dep.isBefore(java.time.Instant.now())) {
          throw new AidError("ERR_INVALID_TXT", "Record is deprecated as of " + depVal);
        }
      } catch (java.time.format.DateTimeParseException e) {
        throw new AidError("ERR_INVALID_TXT", "dep MUST be an ISO 8601 UTC timestamp (e.g., 2026-01-01T00:00:00Z)");
      }
    }

    String uri = hasUri ? raw.get("uri") : raw.get("u");
    if ("local".equals(protoValue)) {
      // must be allowed local scheme
      String scheme = extractScheme(uri);
      if (!isAllowedLocalScheme(scheme)) {
        throw new AidError(
            "ERR_INVALID_TXT",
            "Invalid URI scheme for local protocol. Must be one of: " + String.join(", ", Constants.LOCAL_URI_SCHEMES));
      }
    } else if ("zeroconf".equals(protoValue)) {
      if (!uri.startsWith("zeroconf:")) {
        throw new AidError(
            "ERR_INVALID_TXT",
            "Invalid URI scheme for 'zeroconf'. MUST be 'zeroconf:'");
      }
    } else if ("websocket".equals(protoValue)) {
      if (!uri.startsWith("wss://")) {
        throw new AidError(
            "ERR_INVALID_TXT",
            "Invalid URI scheme for 'websocket'. MUST be 'wss:'");
      }
      try {
        java.net.URI wu = java.net.URI.create(uri);
        if (!"wss".equals(wu.getScheme()) || wu.getHost() == null) {
          throw new IllegalArgumentException();
        }
      } catch (Exception e) {
        throw new AidError("ERR_INVALID_TXT", "Invalid URI format: " + uri);
      }
    } else {
      if (!uri.startsWith("https://")) {
        throw new AidError(
            "ERR_INVALID_TXT",
            "Invalid URI scheme for remote protocol '" + protoValue + "'. MUST be 'https:'");
      }
      try {
        URI u = URI.create(uri);
        if (u.getScheme() == null || u.getHost() == null) {
          throw new IllegalArgumentException();
        }
      } catch (Exception e) {
        throw new AidError("ERR_INVALID_TXT", "Invalid URI format: " + uri);
      }
    }

    String auth = authVal;
    String desc = descVal;
    if (raw.containsKey("pka") && raw.containsKey("k")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"pka\" and \"k\" fields");
    }
    if (raw.containsKey("kid") && raw.containsKey("i")) {
      throw new AidError("ERR_INVALID_TXT", "Cannot specify both \"kid\" and \"i\" fields");
    }
    String pkaVal = raw.containsKey("pka") ? raw.get("pka") : (raw.containsKey("k") ? raw.get("k") : null);
    String kidVal = raw.containsKey("kid") ? raw.get("kid") : (raw.containsKey("i") ? raw.get("i") : null);
    if (pkaVal != null && kidVal == null) {
      throw new AidError("ERR_INVALID_TXT", "kid is required when pka is present");
    }
    return new AidRecord(Constants.SPEC_VERSION, uri, protoValue, auth, desc, docsVal, depVal, pkaVal, kidVal);
  }

  public static boolean isValidProto(String token) {
    // Compare against generated PROTO_* constants
    return Constants.PROTO_MCP.equals(token)
        || Constants.PROTO_A2A.equals(token)
        || Constants.PROTO_OPENAPI.equals(token)
        || Constants.PROTO_LOCAL.equals(token)
        || Constants.PROTO_GRPC.equals(token)
        || Constants.PROTO_GRAPHQL.equals(token)
        || Constants.PROTO_WEBSOCKET.equals(token)
        || Constants.PROTO_ZEROCONF.equals(token);
  }

  private static boolean isValidAuth(String token) {
    return "none".equals(token)
        || "pat".equals(token)
        || "apikey".equals(token)
        || "basic".equals(token)
        || "oauth2_device".equals(token)
        || "oauth2_code".equals(token)
        || "mtls".equals(token)
        || "custom".equals(token);
  }

  private static boolean isAllowedLocalScheme(String scheme) {
    if (scheme == null) return false;
    for (String s : Constants.LOCAL_URI_SCHEMES) {
      if (s.equals(scheme)) return true;
    }
    return false;
  }

  private static String extractScheme(String uri) {
    int idx = uri.indexOf(':');
    if (idx <= 0) return null;
    return uri.substring(0, idx);
  }
}

File: packages/aid-rs/src/parser.rs
use std::collections::HashSet;

use crate::constants_gen::{
    AUTH_APIKEY, AUTH_BASIC, AUTH_CUSTOM, AUTH_MTLS, AUTH_NONE, AUTH_OAUTH2_CODE, AUTH_OAUTH2_DEVICE,
    AUTH_PAT, PROTO_A2A, PROTO_GRAPHQL, PROTO_GRPC, PROTO_LOCAL, PROTO_MCP, PROTO_OPENAPI,
    PROTO_WEBSOCKET, PROTO_ZEROCONF, SPEC_VERSION,
};
use crate::errors::AidError;
use crate::record::AidRecord;

fn is_supported_proto(token: &str) -> bool {
    matches!(
        token,
        PROTO_MCP | PROTO_A2A | PROTO_OPENAPI | PROTO_LOCAL | PROTO_GRPC | PROTO_GRAPHQL | PROTO_WEBSOCKET | PROTO_ZEROCONF
    )
}

fn is_supported_auth(token: &str) -> bool {
    matches!(
        token,
        AUTH_NONE | AUTH_PAT | AUTH_APIKEY | AUTH_BASIC | AUTH_OAUTH2_DEVICE | AUTH_OAUTH2_CODE | AUTH_MTLS | AUTH_CUSTOM
    )
}

pub fn parse(txt: &str) -> Result<AidRecord, AidError> {
    let mut v: Option<String> = None;
    let mut uri: Option<String> = None;
    let mut proto: Option<String> = None;
    let mut p: Option<String> = None;
    let mut auth: Option<String> = None;
    let mut desc: Option<String> = None;
    let mut docs: Option<String> = None;
    let mut dep: Option<String> = None;
    let mut pka: Option<String> = None;
    let mut kid: Option<String> = None;

    let mut seen: HashSet<String> = HashSet::new();

    for raw_pair in txt.split(';') {
        let pair = raw_pair.trim();
        if pair.is_empty() { continue; }
        let mut iter = pair.splitn(2, '=');
        let key_raw = iter.next().ok_or_else(|| AidError::invalid_txt("Invalid key-value pair"))?;
        let value_raw = iter.next().ok_or_else(|| AidError::invalid_txt(format!("Invalid key-value pair: {}", pair)))?;
        let key = key_raw.trim().to_lowercase();
        let value = value_raw.trim().to_string();
        if key.is_empty() || value.is_empty() {
            return Err(AidError::invalid_txt(format!("Empty key or value in pair: {}", pair)));
        }
        match key.as_str() {
            "v" | "uri" | "u" | "proto" | "p" | "auth" | "a" | "desc" | "s" | "docs" | "d" | "dep" | "e" | "pka" | "k" | "kid" | "i" => {
                if seen.contains(&key) { return Err(AidError::invalid_txt(format!("Duplicate key: {}", key))); }
                seen.insert(key.clone());
            }
            _ => {}
        }
        match key.as_str() {
            "v" => v = Some(value),
            "uri" | "u" => {
                if uri.is_none() { uri = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"uri\" and \"u\"")); }
            }
            "proto" => proto = Some(value),
            "p" => p = Some(value),
            "auth" | "a" => {
                if auth.is_none() { auth = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"auth\" and \"a\"")); }
            }
            "desc" | "s" => {
                if desc.is_none() { desc = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"desc\" and \"s\"")); }
            }
            "docs" | "d" => {
                if docs.is_none() { docs = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"docs\" and \"d\"")); }
            }
            "dep" | "e" => {
                if dep.is_none() { dep = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"dep\" and \"e\"")); }
            }
            "pka" | "k" => {
                if pka.is_none() { pka = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"pka\" and \"k\"")); }
            }
            "kid" | "i" => {
                if kid.is_none() { kid = Some(value) } else { return Err(AidError::invalid_txt("Cannot specify both \"kid\" and \"i\"")); }
            }
            _ => {}
        }
    }

    let v = v.ok_or_else(|| AidError::invalid_txt("Missing required field: v"))?;
    if v != SPEC_VERSION {
        return Err(AidError::invalid_txt(format!("Unsupported version: {}. Expected: {}", v, SPEC_VERSION)));
    }

    let uri = uri.ok_or_else(|| AidError::invalid_txt("Missing required field: uri"))?;

    if proto.is_some() && p.is_some() { return Err(AidError::invalid_txt("Cannot specify both \"proto\" and \"p\" fields")); }
    if proto.is_none() && p.is_none() { return Err(AidError::invalid_txt("Missing required field: proto (or p)")); }

    let proto_value = proto.or(p).unwrap();

    if !is_supported_proto(&proto_value) { return Err(AidError::unsupported_proto(format!("Unsupported protocol: {}", proto_value))); }

    if let Some(ref auth_val) = auth {
        if !is_supported_auth(auth_val.as_str()) {
            return Err(AidError::invalid_txt(format!("Invalid auth token: {}", auth_val)));
        }
    }

    if let Some(ref d) = desc {
        if d.as_bytes().len() > 60 {
            return Err(AidError::invalid_txt("Description field must be ≤ 60 UTF-8 bytes"));
        }
    }

    // docs must be https URL when present
    if let Some(ref dv) = docs {
        if !dv.starts_with("https://") {
            return Err(AidError::invalid_txt("docs MUST be an absolute https:// URL"));
        }
        // Basic check: must contain host separator after scheme
        if !dv[8..].contains('/') && !dv[8..].contains(':') {
            return Err(AidError::invalid_txt("Invalid docs URL"));
        }
    }

    // dep must end with Z (basic check)
    if let Some(ref dp) = dep {
        if !dp.ends_with('Z') {
            return Err(AidError::invalid_txt("dep MUST be an ISO 8601 UTC timestamp (e.g., 2026-01-01T00:00:00Z)"));
        }
    }

    // URI validation based on protocol
    if proto_value == PROTO_LOCAL {
        // local scheme already minimally validated by consumers in other languages; here we accept
    } else if proto_value == PROTO_ZEROCONF {
        if !uri.starts_with("zeroconf:") { return Err(AidError::invalid_txt("Invalid URI scheme for 'zeroconf'. MUST be 'zeroconf:'")); }
    } else if proto_value == PROTO_WEBSOCKET {
        if !uri.starts_with("wss://") { return Err(AidError::invalid_txt("Invalid URI scheme for 'websocket'. MUST be 'wss:'")); }
    } else {
        if !uri.starts_with("https://") { return Err(AidError::invalid_txt(format!("Invalid URI scheme for remote protocol '{}'. MUST be 'https:'", proto_value))); }
    }

    Ok(AidRecord { v, uri, proto: proto_value, auth, desc, docs, dep, pka, kid })
}
